### DISCUSSION TITLE
How to change dependency from method level to class constructor level?

### DISCUSSION THREAD

**Discussion by sergeda** [zio/zio#4529] [Q&A] [answered]
Method "log" depends on Clock. I'm trying to provide clock once on class creation instead of passing it to function each time. But code doesn't compile. How to do this correctly?

    
    object logger {
      case class LoggerConfig(host: String, port: Int)
  
    type Logger = Has[Logger.Service]
  
    object Logger {
      trait Service {
        def log(message: String): Task[Unit]
      }

    def live: ZLayer[Has[LoggerConfig] with Has[Clock], Nothing, Has[Logger.Service]] = ZLayer.fromServices((config: LoggerConfig, clock: Clock) =>
      new RetryLogger(config, clock)
    )

    class RetryLogger(config: LoggerConfig, clock: Clock) extends Service {
      private val schedule = {
        import zio.duration._
        Schedule.fibonacci(1.second) && Schedule.recurs(5)
      }

      private def post(message: String): Task[Unit] = ???

      def log(message: String): Task[Unit] =
        post(message).retry(schedule).provide(clock)
    }

    }}

**Comment by adamgfraser** [comment, marked as answer]
@sargeda See example below. I think you were just mixing up the service and the module in a few places.

```scala
import zio._
import zio.clock._

object logger {
  case class LoggerConfig(host: String, port: Int)

  type Logger = Has[Logger.Service]

  object Logger {
    trait Service {
      def log(message: String): Task[Unit]
    }

    def live: ZLayer[Has[LoggerConfig] with Clock, Nothing, Logger] =
      ZLayer.fromServices[LoggerConfig, Clock.Service, Logger.Service](
        (config, clock) => new RetryLogger(config, clock)
      )

    class RetryLogger(config: LoggerConfig, clock: Clock.Service)
        extends Service {
      private val schedule = {
        import zio.duration._
        Schedule.fibonacci(1.second) && Schedule.recurs(5)
      }

      private def post(message: String): Task[Unit] = ???

      def log(message: String): Task[Unit] =
        post(message).retry(schedule).provide(Has(clock))
    }

  }
}
```
